### DISCUSSION TITLE
Loading CJS modules through an ES module library

### DISCUSSION THREAD

**Discussion by jerryjappinen** [nuxt/nuxt#16616] [Questions]
I have a utility library that I use for multiple Nuxt 3 apps.

- This library is written in ES6
- It does _not_ have `"type": "module"` in its `package.json`
- I currently transpile it in each Nuxt app (it won't work if I don't)

This utility library has many dependencies written in CJS which cause an error when loaded in my Nuxt app. These dependencies work when I load them in a Nuxt app directly, but when referenced in the utility library, they get loaded differently and throw errors. Nuxt transpiles my code as ESM but fails when a dependency is written in CJS, even though they are different packages.

Example libraries include `contentful.js` and `graphql-request` (so pretty common and widely used libraries that are beyond my reach to make changes in, at least in the near future).

With Nuxt 3's automagic ES/CJS module loading logic, what would be the recommended way to deal with this scenario?

- Is there a way to get consistent module loading behavior regardless of where a dependency is loaded?
- Should I try to build my utility library with Rollup (and if so, what should the output be)?

![image](https://user-images.githubusercontent.com/1101002/195990933-d44fd549-498c-4d7a-bb3f-378e020c81df.png)

**Comment by jerryjappinen** [comment]
I've been working on this further, and it looks like moving these dependencies to `peerDependencies` and having them installed on the Nuxt app also does not help.

So, if I have a `.vue` component in a utility library that has a dependency to something like `@graphcms/rich-text-html-renderer`, it does not seem to work if I set the utility library to transpile (even though it works if that component is stored directly in the Nuxt app).

If anyone can share what the recommended setup is in this use case, I would love to know.

**Comment by marr** [comment]
@jerryjappinen I know this is an old discussion, but were you ever able to get some clarity around this? I have a similar question.
One thing I found is that setting the `vite.optimizeDeps.include` array to include your package may get around some issues.

**Reply by jerryjappinen** [reply]
Hey @marr - Not really. I've kinda given up on this as dealing with this transpiling and module loading stuff takes out all the joy from development. I do it for fun but if it's not fun, I don't do it.

Thanks for the hint though. I'll try it whenever I run into this again.

Godspeed 🙏
